src/cairo/bootloader.md

```diff
@@ -1 +1,199 @@
+| bitwise     |       |     O     |     |            O            |    O     |          O           |
+| ecdsa       |       |           |  O  |                         |    O     |          O           |
+| poseidon    |       |           |     |            O            |    O     |          O           |
+| ec_op       |       |           |     |                         |    O     |          O           |
+| keccak      |       |           |     |                         |          |          O           |
+
+### Hints
+
+Hints are non-deterministic computation (i.e. code that is run outside of the VM to populate values that are validated inside the VM)
+
+For example, when computing the square root of an integer in a prime field, directly computing this using constraints is expensive. So instead, we can use hint to calculate the square root and create a constraint inside the VM that checks that the power of the given square root is equal to the original value.
+
+So given a computation $x=\sqrt{y}$, compute $y$ using a hint and create a constraint: $y=x*x$
+
+$x$ from the perspective of the VM is a wild guess, but the constraint makes sure that it’s valid.
+
+In Cairo 0 language, you can run a hint by embedding a block of Python code. Below is an example of computing a square root of 25.
+
+```python
+[ap] = 25, ap++;
+%{
+    import math
+    memory[ap] = int(math.sqrt(memory[ap - 1]))
+%}
+[ap - 1] = [ap] * [ap], ap++;
+```
+
+When compiling this Cairo 0 program into CASM, the compiler stores this Python code as a string and specifies that this hint should be run before running the ensuing instruction. In this case, the hint should be run before the `[ap - 1] = [ap] * [ap], ap++;` instruction.
+
+## How the "simple bootloader" works
+
+Coming back to the bootloader, as per the previous section, running the bootloader program will result in a memory vector and an execution trace. But since the bootloader program runs Cairo programs inside it, the generated memory and execution trace should include those of the Cairo programs as well. Below is an example of what the memory layout looks like after running the bootloader:
+
+<div style="text-align: center;">
+    <img src="simple_bootloader_memory_layout.png" alt="Simple bootloader memory layout" width="80%">
+</div>
+
+(Note: a **task** here refers to a single Cairo program and its inputs)
+
+In other words, what the bootloader program does is it makes sure that the inner Cairo programs can write to various segments in the memory without any conflicts with each other or with the outer bootloader program.
+
+This is a "simple" use of the bootloader and we can use this functionality using the `simple_bootloader.cairo` file ([link](https://github.com/starkware-libs/cairo-lang/blob/v0.13.1/src/starkware/cairo/bootloaders/simple_bootloader/simple_bootloader.cairo)) in the Cairo repo. We will discuss an "advanced" use of the bootloader in the next section.
+
```

Comment:
It also verifies the byte code of the program as well no? i.e. hashes the program and exports the hash?